src/dataflow-types/src/types.rs

```diff
@@ -825,19 +873,119 @@ impl ExternalSourceConnector {
     ///
     /// The columns declared here must be kept in sync with the actual source
     /// implementations that produce these columns.
-    pub fn metadata_columns(&self) -> Vec<(ColumnName, ColumnType)> {
+    pub fn metadata_columns<'a>(&'a self, include_defaults: bool) -> Vec<(&'a str, ColumnType)> {
+        let mut columns = Vec::new();
+        let default_col = |name| (name, ScalarType::Int64.nullable(false));
         match self {
-            Self::Kafka(_) => vec![("mz_offset".into(), ScalarType::Int64.nullable(false))],
-            Self::File(_) => vec![("mz_line_no".into(), ScalarType::Int64.nullable(false))],
-            Self::Kinesis(_) => vec![("mz_offset".into(), ScalarType::Int64.nullable(false))],
-            Self::AvroOcf(_) => vec![("mz_obj_no".into(), ScalarType::Int64.nullable(false))],
+            Self::Kafka(KafkaSourceConnector {
+                include_partition: part,
+                include_timestamp: time,
+                include_topic: topic,
+                include_offset: offset,
+                ..
+            }) => {
+                let mut items = BTreeMap::new();
+                // put the offset at the end if necessary
+                if include_defaults && offset.is_none() {
+                    items.insert(4, default_col("mz_offset"));
+                }
+
+                for (include, ty) in [
+                    (offset, ScalarType::Int64),
```

Comment:
That only works if `include_defaults` is true, which... probably we want to eventually become false by default, possibly not at the same time as we entirely remove the code.

Also, I can't get it to compile because of the kinds of references involved.